### ISSUE TITLE
clang mistakenly invokes clang-cl driver when the executable name ends with "cl" or contains "cl-"

### ISSUE THREAD

**Issue by silverweed** [llvm/llvm-project#225385]
Invoking clang with an executable name that contains `cl-` or ends with `cl` causes clang to invoke `clang-cl`, causing unexpected errors.

**Clang version**: 23.1.1
**Platform**: reproduced on Linux and MacOS (didn't try windows)

**Minimal reproducer**:
- rename `clang` to `cl-ang` (or `cl`, or `whateveryouwantcl`)
- try to compile any C or C++ program
- clang reports:
```
cl-ang: warning: unable to find a Visual Studio installation; try running Clang from a developer command prompt [-Wmsvc-not-found]
cl-ang: error: unable to execute command: posix_spawn failed: No such file or directory
cl-ang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Less minimal but more relevant reproducer**:
[The ROOT framework](https://github.com/root-project/root) uses Clang internally to power Cling, its C++ interpreter. When ROOT is initialized, it also initializes Cling, a complex process that involves invoking Clang with some predefined set of flags. Crucially, when Clang is invoked this way, the `ClangExecutable` argument (used in `Driver::BuildCompilation`, which is where this bug likely originates) is not set to `clang` but to the name of the executable that is initializing ROOT.

So, without changing the name of the `clang` binary, one can easily reproduce this behavior by compiling this ROOT code:
```
// compile with: c++ -o cl thisfile.cpp $(root-config --cflags --libs)  (you need to source thisroot.sh) 
#include <TROOT.h>
int main() {
   // this causes ROOT to be initialized
   gROOT;
}
```
and name the resulting binary `cl` (or anything containing `cl-` or ending with `cl`). In this case, the errors will come from Cling, in a form similar to this:
```
warning: unknown argument ignored in clang-cl: '-fcxx-modules'
warning: unknown argument ignored in clang-cl: '-fmodules-cache-path=/home/user/root_build/debug/lib'
warning: unknown argument ignored in clang-cl: '-std=c++17'
warning: unknown argument ignored in clang-cl: '-fsigned-char'
warning: unknown argument ignored in clang-cl: '-fsized-deallocation'
warning: unknown argument ignored in clang-cl: '-fmodules-cache-path=/home/user/root_build/debug/lib'
warning: unknown argument ignored in clang-cl: '-fno-implicit-module-maps'
warning: unknown argument ignored in clang-cl: '-Rmodule-build'
warning: unknown argument ignored in clang-cl: '-fno-autolink'
warning: unknown argument ignored in clang-cl: '-cxx-isystem'
warning: unknown argument ignored in clang-cl: '-cxx-isystem'
warning: unknown argument ignored in clang-cl: '-cxx-isystem'
warning: unknown argument ignored in clang-cl: '-nostdinc++'
warning: unable to find a Visual Studio installation; try running Clang from a developer command prompt
cl: /home/user/root/interpreter/cling/lib/Interpreter/CIFactory.cpp:1025: bool (anonymous namespace)::SetupCompiler(CompilerInstance *, const CompilerOptions &, bool, bool): Assertion `LangOpts.CPlusPlus17 && "Language version mismatch"' failed.
fish: Job 1, './cl' terminated by signal SIGABRT (Abort)
```

**Comment by efriedma-quic** [comment]
The behavior is intentional for the clang executable itself: it lets us symlink clang++/clang-cl/clang-cpp together.  And there's no good reason to go in and rename the executables yourself.

Not sure about the library interface; maybe some tweak to the APIs is appropriate.

**Comment by Andarwinux** [comment]
as a workaround, add `--driver-mode=g++`

**Comment by silverweed** [comment]
> The behavior is intentional for the clang executable itself

The surprising part for me is how easy it is to "trick" clang to execute that behavior, in that it looks very loosely for the presence of `cl` in the exe name; does this come from the standard llvm option parsing logic? Or is the check intentionally this loose? (I'm not familiar with clang-cl so I don't know which patterns are supposed to trigger this behavior)

**Comment by efriedma-quic** [comment]
The logic is in ToolChain::getTargetAndModeFromProgramName().  The underlying assumption is that the executable name should have one of a few canonical forms.

**Comment by silverweed** [comment]
If the behavior of clang itself is intentional this issue can be closed; we are going to fix the clang invocation on our side regardless.
Thank you for the help.
